### PR TITLE
CDAP-18305 support BIGINT properly

### PR DESCRIPTION
**Pull request by seanzhougoogle**

### CODE REVIEW

File: database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java [hunk 1]
```diff
@@ -78,8 +78,14 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
         break;
 
       case Types.BIGINT:
-        type = Schema.Type.LONG;
-        break;
+        //SQL BIGINT is 64 bit, thus signed can be stored in long without loosing precision
```
Comment (avirmani) [inline database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java]: loosing -> losing

Comment (seanzhougoogle) [reply database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java]: done :)

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java [hunk 2]
```diff
@@ -150,6 +151,15 @@ protected void setField(ResultSet resultSet, StructuredRecord.Builder recordBuil
...
     } else if (o instanceof BigDecimal) {
       recordBuilder.setDecimal(field.getName(), (BigDecimal) o);
+    } else if (o instanceof BigInteger) {
+      if (sqlType == Types.BIGINT && (resultSet.getMetaData().isSigned(columnIndex) ||
+          resultSet.getMetaData().getPrecision(columnIndex) < 19)) {
+        //SQL BIGINT type is 64-bit long thus signed should be able to convert to long without loosing precisions
+        // or UNSIGNED type is within the scope of signed long
+        recordBuilder.set(field.getName(), ((BigInteger) o).longValueExact());
+      } else {
+        recordBuilder.setDecimal(field.getName(), new BigDecimal((BigInteger) o, 0));
```
Comment (yaojiefeng) [inline database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:162]: why this scale is 0?

File: database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java [hunk 3]
```diff
@@ -78,8 +78,14 @@ public Schema getSchema(ResultSetMetaData metadata, int index) throws SQLExcepti
...
       case Types.BIGINT:
-        type = Schema.Type.LONG;
-        break;
+        //SQL BIGINT is 64 bit, thus signed can be stored in long without losing precision
+        //or unsigned BIGINT is within the scope of signed long
+        if (metadata.isSigned(index) || metadata.getPrecision(index) < 19) {
+          type = Schema.Type.LONG;
+          break;
+        } else {
+          return Schema.decimalOf(metadata.getPrecision(index));
```
Comment (yaojiefeng) [inline database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java:88]: do we need scale here?

Comment (seanzhougoogle) [reply database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java:88]: default value is 0.

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 2 above)
Comment (seanzhougoogle) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:162]: scale is 0 for a big integer.

**Review by yaojiefeng — APPROVED** [review]

File: database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java (hunk 2 above)
Comment (seanzhougoogle) [reply database-commons/src/main/java/io/cdap/plugin/db/DBRecord.java:162]: add more comments here :)

File: database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java (hunk 3 above)
Comment (seanzhougoogle) [reply database-commons/src/main/java/io/cdap/plugin/db/CommonSchemaReader.java:88]: added more comments here.

**Pull request merged**